style.md

```diff
@@ -0,0 +1,281 @@
+    y[i] = x[i] + 2
+end
+y
+
+# GOOD! #No way to preallocate y (although easier ways to write)
+x = rand(10)
+y = similar(x, 0) #Empty of same type as x
+for val in x
+    if val < 0.5
+        push!(y, val)
+    end
+end
+y
+```
+- **Use iterators directly** rather than accessing by index.
+```julia
+x = [1, 2, 3]
+
+# BAD!
+n = 0
+for i in eachindex(x)
+  n += x[i]^2
+end
+n
+
+# BETTER!
+n = 0
+for val in x
+  n += val^2
+end
+n
+
+# BEST!
+sum(xval -> xval^2, x) #i.e. transform each x and then reduce
+```
+
+- **Avoid `range` when possible** and use the `1.0:0.1:1.0` style notation, etc.  Getting rid of the `linspace` is a pain, but use ranges if the exact top doesn't matter.
+```julia
+Base.OneTo(5) # BAD!
+range(1, stop=5) #BAD!
+1:5 #GOOD!
+
+
+N = 10
+zmin = 0.0
+zstop = 1.0
+
+# Old linspace(zmin, zstop, N) doesn't work
+# ACCEPTABLE!
+range(zmin, stop = zstop, length = N)
+```    
+  - Furthermore, if you don't really care if if hits the `zstop` exactly and are willing to give a stepsize then, the following is the clearest
+```julia
+zmin = 0.0
+zstop = 1.0
+step = 0.1
+
+# GOOD! But..
+r = zmin:step:zstop
+
+# ...CAREFUL
+r = 0.0:0.22:1.0 # Note the end isn't a multiple of the step...
+@assert r == 0.0:0.22:0.88
+@assert maximum(r) == 0.88 #Use to get the maxium of the range, perhaps != 
+```
+
+
+
+## Dependencies
+
+- **Use external packages** whenever possible, and never rewrite code that is available in a well-maintained external package (even if it is imperfect)
+- The following packages can be used as a dependency without any concerns: `QuantEcon, Parameters, Optim, Roots, Expectations, NLsolve, DataFrames, LaTeXStrings, Plots, ....`
+- **Do** use `using` where possible (i.e. not `import`), and include the whole package as opposed to selecting only particular functions or types.
+- **Prefer** to keep packages used throughout the lecture at the top of the first block (e.g. `using LinearAlgebra, Parameters, QuantEcon`)  but packages used only in a single place should have the `using` local to that use.
+    - If `Plots` is only used lower down in the lecture, then try to have it local to that section to ensure faster loading time.
+- **Always seed random numbers** in order for automated testing to function.
+
+## Work in Progress Discussions
+1. When and how do we use `eachindex` with multidimensional arrays you want to iterate through?
+2. How best to stack arrays and unpack them for use with solvers/etc.?  
+3. Simple error handling of reasonable failures (e.g. returning a union with `Nothing`, etc.)  I am not sure we want to teach them about exceptions, for example.
+4. Can we really use Latex for labels?  Is it stable enough with various backends?  If so, I think that LatexStrings is the way to go.
+5. Lets say `x` is vector of length `N`.  Is `y = similar(x, N)` or `y = similar(x)` a better way to denote the preallocation of the same size?  I kind of like the first in this case, but I think we can use our judgement.
```

Comment:
A few other things which come to mind: 

1. A line about using `Missing` from Julia Base, instead of `"N/A"`, `NaN`, `Inf`, or something like that. 

2. A line saying don't use the ternary operator (or avoid it if you can), since I know you don't like it. 

3. A line about meaningful names for variables and functions.

Reply:
The `Missing` and `Nothing` is covered on the Julia style guide.

Reply:
Exactly.  No need to repeat those sorts of things.  The only reason to add something on that for this code is if we find an error handling pattern we think is simple and teachable, but lets wait on that.

Reply:
Gotcha.